google-cloud-pubsub/lib/google/cloud/pubsub/subscriber.rb

```diff
@@ -0,0 +1,386 @@
+          deadline_ack_request.modify_deadline_ack_ids += deadline_ack_ids
+          deadline_ack_request.modify_deadline_seconds += deadline_seconds
+
+          synchronize do
+            @request_queue.push deadline_ack_request
+            @inventory.remove deadline_ack_ids
+            unpause_streaming!
+          end
+
+          true
+        end
+
+        def inventory
+          synchronize { @inventory.count }
+        end
+
+        ##
+        # @private
+        def delay_inventory!
+          synchronize do
+            return true if @inventory.empty?
+
+            ack_ids_deadline = @inventory.ack_ids.map { deadline }
+            inventory_ack_request = Google::Pubsub::V1::StreamingPullRequest.new
+            inventory_ack_request.modify_deadline_ack_ids += @inventory.ack_ids
+            inventory_ack_request.modify_deadline_seconds += ack_ids_deadline
+
+            @request_queue.push inventory_ack_request
+          end
+
+          true
+        end
+
+        ##
+        # @private
+        def clear_inventory!
+          synchronize do
+            @inventory.clear
+            unpause_streaming!
+          end
+        end
+
+        # @private
+        def to_s
+          format "(subscription: %s, inventory: %i)", subscription_name,
+                 inventory
+        end
+
+        # @private
+        def inspect
+          "#<#{self.class.name} #{self}>"
+        end
+
+        protected
+
+        # rubocop:disable all
+
+        def background_run
+          until synchronize { @stopped }
+            Thread.current.kill if synchronize { @paused }
+
+            begin
+              # Cannot syncronize the enumerator, causes deadlock
+              response = @output_enum.next
+              response.received_messages.each do |rec_msg_grpc|
+                rec_msg = ReceivedMessage.from_grpc(rec_msg_grpc, self)
+                synchronize do
+                  @inventory.add rec_msg.ack_id
+
+                  perform_callback_async rec_msg
+                end
+              end
+              synchronize { pause_streaming! }
+            rescue StopIteration
+              break
+            end
+          end
+        rescue GRPC::DeadlineExceeded
+          # The GRPC client will raise when stream is opened longer than the
+          # timeout value it is configured for. When this happends, restart the
+          # stream stealthly.
+        rescue => e
+          synchronize do
+            @request_queue.unshift Google::Cloud::Error.from_error(e)
+          end
+        ensure
+          Thread.pass
+        end
+
+        # rubocop:enable all
+
+        def perform_callback_async rec_msg
+          Concurrent::Future.new(executor: @thread_pool) do
+            @callback.call rec_msg
+          end.execute
+        end
+
+        def start_streaming!
+          @request_queue.unshift initial_input_request
+          @output_enum = service.streaming_pull @request_queue.each_item
```

Comment:
I meant "adding to the initial request being more successful" is an implementation detail, not this function.

If it's larger than 10 MB then we can only guarantee the first 10 MB of acks will be sent successfully. The others should be sent on a small delay (0.5s) after which the stream should have completed connecting.